### PR TITLE
Combine and enhance the two existing OpenAI integrations

### PR DESCRIPTION
**Pull request by ali-sawyer**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

Ready to start review/discussion. However, I am still trying to figure out docker stuff. This integration would require a new docker image, because it requires python libraries that are not currently available on any of the demisto images. I am trying to contribute the new image to https://github.com/demisto/dockerfiles, but have not contributed to this repo before and may need some help.

## Related Issues
N/A

## Description
This PR would combine the two existing, community-supported OpenAI integrations into one to simplify the user experience, and make the integrations easier to update and maintain going forward.

It would also add the following features:
- Add support for Azure OpenAI (using the work of Tim Roberts, want to make sure he gets credit)
- Add two new commands optimized to support two common generative AI use cases:  1.) question-answering and 2.) summarization
- Through the two new commands mentioned above, add support for prompts longer than 4097 characters through use of LangChain embeddings 
- Add capability to "remember" chat history via the question-answering command

I am of the opinion that we should combine standard OpenAI and Azure OpenAI into a single integration, to simplify future updates and help ensure we are keeping both services in sync. Per Microsoft:
"Azure OpenAI co-develops the APIs with OpenAI, ensuring compatibility and a smooth transition from one to the other."
https://learn.microsoft.com/en-us/azure/cognitive-services/openai/faq#how-do-the-capabilities-of-azure-openai-compare-to-openai-

That being said, I am open to discussion and happy to make any required changes to the proposed integration.

## Screenshots
N/A

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [x] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details: I propose deprecating the two existing integrations so they still exist, and users can continue using them if they wish. Otherwise, if this integration replaces the two existing ones, one command name would change: `chatgpt-send-prompt` would become `openai-chatgpt`.
   - [x] No

## Must have
- [ ] Tests
- [x] Documentation

### CODE REVIEW

**Comment by ali-sawyer** [comment]
dockerfiles repo PR: https://github.com/demisto/dockerfiles/pull/17642

**Comment by anas-yousef** [comment]
It was decided to insert the modifications and new features to the `OpenAI` integration, where it will also include the commands of the `OpenaiChatGPTV3` integration, which will be marked as deprecated.

File: Packs/OpenAI/Integrations/OpenAI/OpenAI.py [hunk 1]
```diff
@@ -123,41 +299,184 @@ def reputations_command(client: Client, args: dict) -> CommandResults:
...
 
-    try:
-        client = Client(
-            base_url=base_url,
-            api_key=api_key,
-            verify=verify,
-            proxy=proxy
-        )
+    # set env variable for API key
+    os.environ["OPENAI_API_KEY"] = api_key
```
Comment (anas-yousef) [inline Packs/OpenAI/Integrations/OpenAI/OpenAI.py:441]: What is the reason of setting env variables? I saw that for the [FAISS](https://python.langchain.com/docs/modules/data_connection/vectorstores/integrations/faiss) library, this is needed, are the other env variables for the Azure OpenAI vendor?

File: Packs/OpenAI/Integrations/OpenAI/OpenAI.yml [hunk 2]
```diff
@@ -1,64 +1,113 @@
...
       - text-curie-001
       - text-babbage-001
       - text-ada-001
       - code-davinci-002
       - code-cushman-001
-      description: The model which will generate the completion. Some models are suitable for natural language tasks, others specialize in code
-      defaultValue: text-davinci-003
-      type: textArea
-    - name: temperature
+    - defaultValue: "0.2"
```
Comment (anas-yousef) [inline Packs/OpenAI/Integrations/OpenAI/OpenAI.yml:89]: Is there a reason this was changed to 0.2?

File: Packs/OpenAI/Integrations/OpenAI/OpenAI.yml [hunk 3]
```diff
@@ -1,64 +1,113 @@
...
-  defaultvalue: https://api.openai.com/
-  type: 0
   required: true
-- display: API Key
+  section: Connect
+  type: 0
+- additionalinfo: OpenAI API key
+  display: API Key
   name: apikey
+  required: false
```
Comment (anas-yousef) [inline Packs/OpenAI/Integrations/OpenAI/OpenAI.yml:15]: Why was this changed to false?

File: Packs/OpenAI/ReleaseNotes/1_2_0.md [hunk 4]
```diff
@@ -0,0 +1,6 @@
+
+#### Integrations
+
+##### New: OpenAI v2
```
Comment (anas-yousef) [inline Packs/OpenAI/ReleaseNotes/1_2_0.md:4]: Update the release notes to reflect your latest changes.

File: Packs/OpenAI/Integrations/OpenAI/README.md [hunk 5]
```diff
@@ -1,63 +1,238 @@
...
+#### Human Readable Output
+
+>**Summary**: This guide provides best practices for optimizing the performance of Cortex XSOAR, a security orchestration, automation, and response platform. It includes configuration best practices, such as enabling signature verification on incoming SAML authentication tokens, restricting workers for all time searches, and database optimizations. Additionally, it provides help on playbook design and best practices, such as using Quiet Mode and disabling auto-extract on tasks that do not require it. It also provides instructions on how to configure a server to restrict the number of workers allowed to execute parallel searches, as well as how to audit index tuning and purge existing items. Additionally, it provides guidance on how to administer, tune, monitor, and maintain Cortex XSOAR for multi-tenant hosts and HA groups.
+
+## Breaking changes from the previous version of this integration - OpenAI
+The following sections list the changes in this version.
+
+### Commands
+#### The following commands were removed in this version:
+* *chatgpt-send-prompt* - this command was replaced by *openai-chatgpt*.
```
Comment (anas-yousef) [inline Packs/OpenAI/Integrations/OpenAI/README.md:232]: How is this breaking changes? The command *chatgpt-send-prompt* was not in the `OpenAI` integration.

File: Packs/OpenAI/Integrations/OpenAI/README.md [hunk 6]
```diff
@@ -1,63 +1,238 @@
...
+            "Created Time": 1687968757,
+            "Model": "gpt-35-turbo",
+            "Number of Completion Tokens": 371,
+            "Number of Prompt Tokens": 101,
+            "Number of Total Tokens": 472,
+            "id": "chatcmpl-7WS4T5dI4r0RR8ZRa19QCM0cERlKt"
```
Comment (anas-yousef) [inline Packs/OpenAI/Integrations/OpenAI/README.md:71]: Is this considered something private? Or can it be used publicly?

File: Packs/OpenAI/Integrations/OpenAI/README.md [hunk 7]
```diff
@@ -1,63 +1,238 @@
...
+>**Summary**: This guide provides best practices for optimizing the performance of Cortex XSOAR, a security orchestration, automation, and response platform. It includes configuration best practices, such as enabling signature verification on incoming SAML authentication tokens, restricting workers for all time searches, and database optimizations. Additionally, it provides help on playbook design and best practices, such as using Quiet Mode and disabling auto-extract on tasks that do not require it. It also provides instructions on how to configure a server to restrict the number of workers allowed to execute parallel searches, as well as how to audit index tuning and purge existing items. Additionally, it provides guidance on how to administer, tune, monitor, and maintain Cortex XSOAR for multi-tenant hosts and HA groups.
+
+## Breaking changes from the previous version of this integration - OpenAI
+The following sections list the changes in this version.
+
+### Commands
+#### The following commands were removed in this version:
+* *chatgpt-send-prompt* - this command was replaced by *openai-chatgpt*.
+
+## Additional Considerations for this version
```
Comment (anas-yousef) [inline Packs/OpenAI/Integrations/OpenAI/README.md:234]: Maybe also add that the `OpenAIChatGPTV3` integration has been combined with the `OpenAI` integration, and the command *chatgpt-send-prompt`* is now the *openai-chatgpt* commmand.

**Comment by anas-yousef** [comment]
Hey @ali-sawyer,
Amazing job with this contribution! I wrote some notes, please go over them, and update me if you have any questions :)

File: Packs/OpenAI/Integrations/OpenAI/README.md (hunk 5 above)
Comment (ali-sawyer) [reply Packs/OpenAI/Integrations/OpenAI/README.md:232]: The command `chatgpt-send-prompt` is in the OOTB `OpenAi ChatGPT v3 (Community Contribution)` integration. So it is a breaking change if the plan is to deprecate `OpenAi ChatGPT v3 (Community Contribution)`.

File: Packs/OpenAI/Integrations/OpenAI/OpenAI.yml (hunk 3 above)
Comment (ali-sawyer) [reply Packs/OpenAI/Integrations/OpenAI/OpenAI.yml:15]: This was unintentional. Fixing.

File: Packs/OpenAI/Integrations/OpenAI/OpenAI.py (hunk 1 above)
Comment (ali-sawyer) [reply Packs/OpenAI/Integrations/OpenAI/OpenAI.py:441]: Yes exactly, for FAISS and for using Azure OpenAI with LangChain: https://python.langchain.com/docs/integrations/providers/azure_openai

File: Packs/OpenAI/Integrations/OpenAI/OpenAI.yml (hunk 2 above)
Comment (ali-sawyer) [reply Packs/OpenAI/Integrations/OpenAI/OpenAI.yml:89]: IMO for most security-related use cases, it is more common to want a deterministic, high-probability result. However, the default value is not a big deal, since it can and should be overridden by the user with the appropriate value for their use case. I am fine with changing the default back to 0.7 for consistency.

File: Packs/OpenAI/Integrations/OpenAI/README.md (hunk 5 above)
Comment (ali-sawyer) [reply Packs/OpenAI/Integrations/OpenAI/README.md:232]: Removed this since the `OpenAi ChatGPT v3 (Community Contribution)` integration will be deprecated separately.

File: Packs/OpenAI/Integrations/OpenAI/README.md (hunk 6 above)
Comment (ali-sawyer) [reply Packs/OpenAI/Integrations/OpenAI/README.md:71]: Nope, it's not private, I don't see any issues with including this. The `id` is just a unique ID for each API response. Both [OpenAI](https://platform.openai.com/docs/api-reference/chat) and [Azure OpenAI](https://learn.microsoft.com/en-us/azure/ai-services/openai/reference) include examples like this in our docs. The number of tokens is just determined from the contents of the prompt/response so it is not sensitive either. 

Anyway I'll change it to a dummy ID `chatcmpl-123` like OpenAI does in their docs.

File: Packs/OpenAI/Integrations/OpenAI/OpenAI.py [hunk 8]
```diff
@@ -1,109 +1,285 @@
...
+        Args:
+            entry_id (str): XSOAR entry ID of file to load
+            text_splitter (RecursiveCharacterTextSplitter): LangChain TextSplitter object
+
+        Returns:
+            list: list of loaded documents
+    """
+    file = demisto.getFilePath(entry_id)
+    file_path = file.get('path')
+    os.rename(file_path, f"{file_path}.txt")
```
[the remark this answers is not in the document]
Comment (ali-sawyer) [reply Packs/OpenAI/Integrations/OpenAI/OpenAI.py:207]: Updated.

**Comment by anas-yousef** [comment]
Closing this contribution.
@ali-sawyer and I discussed it offline, and we saw that we should first create a contribution that will combine both integrations into one, and in another contribution, try to add the [openai](https://pypi.org/project/openai/) package.
